x509/revocation/crl/crl.go

```diff
@@ -0,0 +1,143 @@
+)
+
+type crlNumberExtension struct {
+	ID        asn1.ObjectIdentifier
+	Critical  bool `asn1:"optional"`
+	CRLNumber int
+}
+
+// TODO: handle additional CRL Extensions
+// 2.5.29.21,cRLReason
+// id-ce-cRLReasons OBJECT IDENTIFIER ::= { id-ce 21 }
+// -- reasonCode ::= { CRLReasonCode }
+
+//2.5.29.24,invalidityDate
+// id-ce-invalidityDate OBJECT IDENTIFIER ::= { id-ce 24 }
+// -- InvalidityDate ::=  GeneralizedTime
+
+// RevocationData - information on whether a certificate has been
+// revoked by a specified CRL, and information on the CRL
+type RevocationData struct {
+	CRLSignatureAlgorithm         x509.SignatureAlgorithm
+	CRLSignatureValue             []byte
+	Version                       int `asn1:"optional,default:0"`
+	Issuer                        pkix.Name
+	ThisUpdate                    time.Time
+	NextUpdate                    time.Time `asn1:"optional"`
+	CRLExtensions                 ListExtensionData
+	UnknownCRLExtensions          []pkix.Extension `asn1:"tag:0,optional,explicit"`
+	IsRevoked                     bool
+	RevocationTime                time.Time
+	CertificateEntryExtensions    RevokedCertExtensionData
+	RawCertificateEntryExtensions []pkix.Extension `asn1:"optional"`
+}
+
+// ListExtensionData - Data from optional, non-critical pkix.CertificateList extensions
+type ListExtensionData struct {
+	CRLNumber int
+	AuthKeyID x509.SubjAuthKeyId `json:"authority_key_id,omitempty"`
+}
+
+// RevokedCertExtensionData - Data from optional, non-critical pkix.RevokedCertificate extensions
+type RevokedCertExtensionData struct {
+	Reason         string `json:"revocation_reason,omitempty"`
+	invalidityDate time.Time
+}
+
+func gatherListExtensionInfo(certList *pkix.CertificateList, ret *RevocationData) {
+	for _, extension := range certList.TBSCertList.Extensions {
+		if extension.Id.Equal(crlNumberExtensionOID) {
+			var ext crlNumberExtension
+			asn1.Unmarshal(extension.Value, &ext.CRLNumber)
+			ret.CRLExtensions.CRLNumber = ext.CRLNumber
+		} else {
+			ret.UnknownCRLExtensions = append(ret.UnknownCRLExtensions, extension)
+		}
+	}
+}
+
+// CheckCRLForCert - parses through a given CRL and to see if a given certificate
+// is present, and returns data on the revocation and CRL in general
+func CheckCRLForCert(certList *pkix.CertificateList, cert *x509.Certificate) (*RevocationData, error) {
+	ret := &RevocationData{
+		CRLSignatureAlgorithm: x509.GetSignatureAlgorithmFromAI(certList.SignatureAlgorithm),
+		CRLSignatureValue:     certList.SignatureValue.Bytes,
+		Version:               certList.TBSCertList.Version,
+		ThisUpdate:            certList.TBSCertList.ThisUpdate,
+		NextUpdate:            certList.TBSCertList.NextUpdate,
+		IsRevoked:             false,
+	}
+	ret.Issuer.FillFromRDNSequence(&certList.TBSCertList.Issuer)
+
+	gatherListExtensionInfo(certList, ret)
+
+	revokedCerts := certList.TBSCertList.RevokedCertificates
+	for i := range revokedCerts {
```

Comment:
Would it make sense to store this internally as a map for quick lookup?

Reply:
Good call, I added an optional argument to this function so that clients can pass in a map for quick lookups https://github.com/zmap/zcrypto/pull/142/commits/914bf40eb7e5a3bcd34362f6d7e65f776a1906b8